packages/msi_windows.go

```diff
@@ -0,0 +1,235 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package packages
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"syscall"
+	"unsafe"
+
+	"github.com/GoogleCloudPlatform/osconfig/clog"
+	ole "github.com/go-ole/go-ole"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	msiInstallArgs = []string{"ACTION=INSTALL", "REBOOT=ReallySuppress"}
+
+	msi                        = windows.NewLazySystemDLL("msi.dll")
+	procMsiOpenPackageExW      = msi.NewProc("MsiOpenPackageExW")
+	procMsiGetProductPropertyW = msi.NewProc("MsiGetProductPropertyW")
+	procMsiQueryProductStateW  = msi.NewProc("MsiQueryProductStateW")
+	procMsiCloseHandle         = msi.NewProc("MsiCloseHandle")
+	procMsiInstallProductW     = msi.NewProc("MsiInstallProductW")
+	procMsiSetInternalUI       = msi.NewProc("MsiSetInternalUI")
+
+	once sync.Once
+)
+
+func init() {
+	MSIExists = true
+}
+
+func setUIMode() {
+	/*
+		INSTALLUILEVEL MsiSetInternalUI(
+		  INSTALLUILEVEL dwUILevel,
+		  HWND           *phWnd
+		);
+	*/
+	const INSTALLUILEVEL_NONE = 2
+	once.Do(func() {
+		procMsiSetInternalUI.Call(
+			uintptr(INSTALLUILEVEL_NONE),
+			0,
+		)
+	})
+}
+
+// https://docs.microsoft.com/en-us/windows/win32/api/msi/nf-msi-msiopenpackageexw
+func msiOpenPackageExW(szPackagePath string, dwOptions uint32) (uintptr, error) {
+	/*
+		UINT MsiOpenPackageExW(
+		  LPCWSTR   szPackagePath,
+		  DWORD     dwOptions,
+		  MSIHANDLE *hProduct
+		);
+	*/
+	var handle int32
+	pHandle := uintptr(unsafe.Pointer(&handle))
+
+	szPackagePathPtr, err := syscall.UTF16PtrFromString(szPackagePath)
+	if err != nil {
+		return 0, fmt.Errorf("error encoding szPackagePath to UTF16: %v", err)
+	}
+
+	ret, _, _ := procMsiOpenPackageExW.Call(
+		uintptr(unsafe.Pointer(szPackagePathPtr)),
+		uintptr(dwOptions),
+		pHandle,
+	)
+	if ret != 0 {
+		return 0, fmt.Errorf("MsiOpenPackageExW error: %s", syscall.Errno(ret))
+	}
+
+	return uintptr(handle), nil
+}
+
+// https://docs.microsoft.com/en-us/windows/win32/api/msi/nf-msi-msiclosehandle
+func msiCloseHandle(handle uintptr) {
+	/*
+		UINT MsiCloseHandle(
+		  MSIHANDLE hAny
+		);
+	*/
+
+	procMsiCloseHandle.Call(handle)
+}
+
+// https://docs.microsoft.com/en-us/windows/win32/api/msi/nf-msi-msigetproductpropertyw
+func msiGetProductPropertyW(handle uintptr, szProperty string) (string, error) {
+	/*
+		UINT MsiGetProductPropertyW(
+		  MSIHANDLE hProduct,
+		  LPCSTR    szProperty,
+		  LPSTR     lpValueBuf,
+		  LPDWORD   pcchValueBuf
+		);
+	*/
+
+	szPropertyPtr, err := syscall.UTF16PtrFromString(szProperty)
+	if err != nil {
+		return "", fmt.Errorf("error encoding szProperty to UTF16: %v", err)
+	}
+
+	size := uint32(128)
+	lpValueBuf := make([]uint16, size)
+
+	ret, _, _ := procMsiGetProductPropertyW.Call(
+		handle,
+		uintptr(unsafe.Pointer(szPropertyPtr)),
+		uintptr(unsafe.Pointer(&lpValueBuf[0])),
+		uintptr(unsafe.Pointer(&size)),
+	)
```

Comment:
Is any cleanup needed for the thing that the `szPropertyPtr` points too? Does `unsafe.Pointer()` handle that for you?

Reply:
The go garbage collector should handle that with no additional work needed, once GC runs these out of scope variables will be cleaned up. I will be running stress tests on all the config code before launch on both Linux and Windows before launch to make sure we don't have any leaks but Go is very good about this, it's the Windows COM apis that have issues (the msi handle is a good example where we could leak).